script/tool/lib/src/common/package_state_utils.dart

```diff
@@ -62,34 +74,131 @@ PackageChangeState checkPackageChangeState(
+// published according to https://dart.dev/tools/pub/package-layout#examples.
+//
+// This is not exhastive; it currently only handles variations we actually have
+// in our repositories.
+bool _isUnpublishedExampleChange(
+    List<String> pathComponents, RepositoryPackage package) {
+  if (pathComponents.first != 'example') {
+    return false;
+  }
+  final List<String> exampleComponents = pathComponents.sublist(1);
+  if (exampleComponents.isEmpty) {
+    return false;
+  }
+
+  final Directory exampleDirectory =
+      package.directory.childDirectory('example');
+
+  // Check for example.md/EXAMPLE.md first, as that has priority. If it's
+  // present, any other example file is unpublished.
+  final bool hasExampleMd =
+      exampleDirectory.childFile('example.md').existsSync() ||
+          exampleDirectory.childFile('EXAMPLE.md').existsSync();
+  if (hasExampleMd) {
+    return !(exampleComponents.length == 1 &&
+        exampleComponents.first.toLowerCase() == 'example.md');
+  }
+
+  // Most packages have an example/lib/main.dart (or occasionally
+  // example/main.dart), so check for that. The other naming variations aren't
+  // currently used.
+  const String mainName = 'main.dart';
+  final bool hasExampleCode =
+      exampleDirectory.childDirectory('lib').childFile(mainName).existsSync() ||
+          exampleDirectory.childFile(mainName).existsSync();
+  if (hasExampleCode) {
+    // If there is an example main, only that example file is published.
+    return !((exampleComponents.length == 1 &&
+            exampleComponents.first == mainName) ||
+        (exampleComponents.length == 2 &&
+            exampleComponents.first == 'lib' &&
+            exampleComponents[1] == mainName));
+  }
+
+  // If there's no example code either, the example README.md, if any, is the
+  // file that will be published.
+  return exampleComponents.first.toLowerCase() != 'readme.md';
+}
+
+// True if the change is only relevant to people working on the plugin.
+Future<bool> _isDevChange(List<String> pathComponents,
+    {GitVersionFinder? git, String? repoPath}) async {
+  return _isTestChange(pathComponents) ||
+      // The top-level "tool" directory is for non-client-facing utility
+      // code, such as test scripts.
+      pathComponents.first == 'tool' ||
+      // Ignoring lints doesn't affect clients.
+      pathComponents.contains('lint-baseline.xml') ||
+      await _isGradleTestDependencyChange(pathComponents,
+          git: git, repoPath: repoPath);
+}
+
+Future<bool> _isGradleTestDependencyChange(List<String> pathComponents,
+    {GitVersionFinder? git, String? repoPath}) async {
+  if (git == null) {
+    return false;
+  }
+  if (pathComponents.last != 'build.gradle') {
+    return false;
+  }
+  final List<String> diff = await git.getDiffContents(targetPath: repoPath);
+  final RegExp changeLine = RegExp(r'[+-] ');
+  final RegExp testDependencyLine =
+      RegExp(r'[+-]\s*(?:androidT|t)estImplementation\s');
```

Comment:
My understanding is that `api` affects clients, and thus when they are rolled in that configuration we need a version changed. The use of `api` for test dependencies *should* be unique to `espresso` (AFAIK) because it is itself a test package.